lib/manageiq/cross_repo/runner/base.rb

```diff
@@ -0,0 +1,102 @@
+require "active_support/core_ext/object/blank"
+require "yaml"
+
+module ManageIQ::CrossRepo
+  class Runner
+    class Base
+      attr_accessor :script_cmd
+
+      def initialize(script_cmd = nil)
+        @script_cmd = script_cmd.presence
+      end
+
+      def build_test_script
+        load_config!
+        build_script
+      end
+
+      private
+
+      def environment_setup_commands
+        commands = []
+
+        if config["node_js"]
+          commands << "source ~/.nvm/nvm.sh"
+          commands += Array(config["node_js"]).map do |node_version|
+            "nvm install #{node_version}"
+          end
+        end
+
+        commands.any? ? build_section("environment", *commands) : commands
+      end
+
+      def section_commands
+        sections = %w[before_install install before_script script]
+        sections.flat_map do |section|
+          commands = build_section_commands(section)
+          build_section(section, *commands) if commands.present?
+        end.compact
+      end
+
+      def build_commands
+        environment_setup_commands + section_commands
+      end
+
+      def build_section_commands(section)
+        # Travis sections can have a single command or an array of commands
+        Array(config[section]).map { |cmd| "#{cmd} || exit $?" }
+      end
+
+      def build_section(section, *commands)
+        [
+          "echo 'travis_fold:start:#{section}'",
+          *commands,
+          "echo 'travis_fold:end:#{section}'"
+        ]
+      end
+
+      def build_script
+        <<~BASH_SCRIPT
+          #!/bin/bash
+
+          #{build_commands.join("\n")}
+        BASH_SCRIPT
+      end
+
+      def load_config!
+        config
+      end
+
+      def config
+        @config ||= travis_config.tap do |config|
+          # Set missing sections to the proper defaults
+          config["install"] ||= defaults[config["language"]]["install"]
+
+          config["script"] = script_cmd if script_cmd.present?
+          config["script"] ||= defaults[config["language"]]["script"]
+        end
+      end
+
+      def travis_config
```

Comment:
`travis_config` is a little confusing since one of the sources is travis but this is implemented by both the travis and github subclasses.

This looks like just setting defaults, maybe have `config` be implemented by the subclasses directly and reverse_merge `default_config` or something?

Reply:
So I named it `travis_config` to say "This is the repo's config in a travis-style, regardless of what the native format is".  Because that's what I want to normalize to at the moment...maybe `travis_style_config`? The name of this was hard.